compute_tools/src/bin/compute_ctl.rs

```diff
@@ -271,6 +271,38 @@ fn main() -> Result<()> {
         }
     };
 
+    // start the vm-monitor if directed to
+    // the vm-monitor only runs on linux because it requires cgroups
+    #[cfg(target_os = "linux")]
+    let vm_monitor = {
+        let vm_monitor_addr = matches.get_one::<String>("vm-monitor-addr");
+        let cgroup = matches.get_one::<String>("filecache-connstr");
+        let file_cache_connstr = matches.get_one::<String>("cgroup");
+        match (env::var_os("AUTOSCALING"), vm_monitor_addr) {
+            (None, None) => None,
+            (None, Some(_)) => {
+                warn!("--vm-monitor-addr option set but AUTOSCALING env var not present");
+                None
+            }
+            (Some(_), None) => {
+                panic!("AUTOSCALING env var present but --vm-monitor-addr option not set")
+            }
+            (Some(_), Some(addr)) => {
+                let rt = tokio::runtime::Builder::new_current_thread()
+                    .enable_all()
+                    .build()
+                    .expect("failed to create rt");
+                Some(
+                    rt.spawn(vm_monitor::start(Box::leak(Box::new(vm_monitor::Args {
+                        cgroup: cgroup.cloned(),
+                        pgconnstr: file_cache_connstr.cloned(),
+                        addr: addr.clone(),
+                    })))),
+                )
```

Comment:
I should had seen this but here it's obvious; the `rt` get's dropped right away.

Spawning a thread and doing:

```
rt.spawn(...);
rt.block_on(futures::future::pending::<()>());
```

Would work, since block_on is the only way runtime will get control.

Otherwise I was really happy to see current thread rt!

Reply:
That's what I though too, but from the docs:
> The thread initiating the shutdown blocks until all spawned work has been stopped. This can take an indefinite amount of time. The Drop implementation waits forever for this.

I don't know if this means that the runtime will still wait for the spawned work, even if `shutdown` is not called.

Reply:
> > The thread initiating the shutdown blocks until all spawned work has been stopped
>
> I don't know if this means that the runtime will still wait for the spawned work

I don't believe it will. IIRC, shutdown will wait for _currently running_ tasks to yield, but will not start executing something new. I may be wrong though, it's been a while.

Reply:
Should be taken care of by this commit
gracefully shutdown monitor with CancellationToken: b50501dbb74b048013b76ea3bd3292dd3696f729